src/server/pps/server/s3g_sidecar.go

```diff
@@ -0,0 +1,482 @@
+		apiServer:    a,
+		pipelineInfo: &pps.PipelineInfo{}, // populate below
+		pachClient:   a.env.GetPachClient(context.Background()),
+		servers:      make(map[string]*http.Server),
+	}
+
+	// Read spec commit for this sidecar's pipeline, and set auth token for pach
+	// client
+	specCommit := a.env.PPSSpecCommitID
+	if specCommit == "" {
+		// This error is not recoverable
+		panic("cannot serve sidecar S3 gateway if no spec commit is set")
+	}
+	if err := backoff.RetryNotify(func() error {
+		retryCtx, retryCancel := context.WithCancel(context.Background())
+		defer retryCancel()
+		if err := a.sudo(s.pachClient.WithCtx(retryCtx), func(superUserClient *client.APIClient) error {
+			s.pipelineInfo.Reset()
+			buf := bytes.Buffer{}
+			if err := superUserClient.GetFile(ppsconsts.SpecRepo, specCommit, ppsconsts.SpecFile, 0, 0, &buf); err != nil {
+				return fmt.Errorf("could not read existing PipelineInfo from PFS: %v", err)
+			}
+			if err := s.pipelineInfo.Unmarshal(buf.Bytes()); err != nil {
+				return fmt.Errorf("could not unmarshal PipelineInfo bytes from PFS: %v", err)
+			}
+			return nil
+		}); err != nil {
+			return fmt.Errorf("sidecar s3 gateway: could not read pipeline spec commit: %v", err)
+		}
+		if !ppsutil.ContainsS3Inputs(s.pipelineInfo.Input) && !s.pipelineInfo.S3Out {
+			return nil // break out of backoff (nothing to serve via S3 gateway)
+		}
+
+		// Set auth token for s.pachClient (pipelinePtr.AuthToken will be empty if
+		// auth is off)
+		pipelineName := s.pipelineInfo.Pipeline.Name
+		pipelinePtr := &pps.EtcdPipelineInfo{}
+		err := a.pipelines.ReadOnly(retryCtx).Get(pipelineName, pipelinePtr)
+		if err != nil {
+			return fmt.Errorf("could not get auth token from etcdPipelineInfo: %v", err)
+		}
+		s.pachClient.SetAuthToken(pipelinePtr.AuthToken)
+		return nil
+	}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
+		logrus.Errorf("error starting sidecar s3 gateway: %v; retrying in %d", err, d)
+		return nil
+	}); err != nil {
+		// This code should never run, but I hesitate to introduce a panic to new
+		// code
+		logrus.Errorf("restarting startup of sidecar s3 gateway: %v", err)
+		a.ServeSidecarS3G()
+	}
+	if !ppsutil.ContainsS3Inputs(s.pipelineInfo.Input) && !s.pipelineInfo.S3Out {
+		return // break early (nothing to serve via S3 gateway)
+	}
+
+	// begin creating k8s services and s3 gateway instances for each job
+	done := make(chan string)
+	go func() {
+		s.createK8sServices()
+		done <- "createK8sServices"
+	}()
+	go func() {
+		s.serveS3Instances()
+		done <- "serveS3Instances"
+	}()
+	finisher := <-done
+	panic(
+		fmt.Sprintf("sidecar s3 gateway: %s is exiting, which should never happen", finisher),
+	)
+}
+
+func (s *sidecarS3G) serveS3Instances() {
+	// Watch for new jobs & initialize s3g for each new job
+	(&handleJobsCtx{
+		s: s,
+		h: &s3InstanceCreatingJobHandler{s},
+	}).start()
+	panic("sidecar s3 gateway: serveS3Instances() is exiting, which shouldn't happen")
```

Comment:
this panic is redundant, since exits are captured by the `done` channel

Reply:
Ah, thanks! Removed.